std/stdio.d

```diff
@@ -4043,34 +4083,38 @@ private struct ReadlnAppender
     import core.stdc.string;
 
     char[] buf;
-    size_t cap;
     size_t pos;
-    size_t initcap;
+    bool newBuf = false;
 
     void initialize(char[] b)
     {
         buf = b;
-        cap = initcap = buf.capacity;
-        if (cap < buf.length)
-            cap = buf.length;
         pos = 0;
     }
     @property char[] data()
     {
-        // always shrink/extend the buffer to allow reuse in the next call
-        if (initcap > 0 || cap > buf.length)
+        /* When buf is not known to the outside world, it's safe to append to
+        the returned slice. */
+        if (newBuf)
             assumeSafeAppend(buf.ptr[0..pos]);
         return buf.ptr[0..pos];
     }
 
     void reserve(size_t n)
     {
-        if (pos + n > cap)
+        if (buf.length >= pos + n) // buf is already large enough
+            return;
+
+        if (buf.capacity >= pos + n)
+        {
+            buf.length = pos + n;
+        }
+        else
         {
-            size_t ncap = cap * 2 + 128 + n;
+            newBuf = true;
+            size_t ncap = buf.length * 2 + 128 + n;
             char[] nbuf = new char[ncap];
             memcpy(nbuf.ptr, buf.ptr, pos);
-            cap = nbuf.capacity;
             buf = nbuf.ptr[0 .. buf.length];   // remember initial length
```

Comment:
This isn't right. The original code ignored buf.length and used cap. You are using buf.length for what cap did. So this essentially allocates no new capacity.

I think you want `buf = nbuf`